core/modules/filter/filter.pages.inc

```diff
@@ -491,28 +491,169 @@ function filter_format_editor_dialog_save($form, &$form_state) {
+  $dangerous_protocols = _filter_check_dangerous_protocols($value);
+
   // Unlike #type = 'url' validation, we allow both relative and absolute paths.
-  if ($value !== '' && !valid_url($value, TRUE) && !valid_url($value, FALSE)) {
+  if ($dangerous_protocols || (!valid_url($value, TRUE) && !valid_url($value, FALSE))) {
     form_error($element, t('The URL %url is not valid.', array('%url' => $value)));
   }
 }
 
+/**
+ * Clean up URL for validation.
+ *
+ * Attempts to create a URL that will pass validation: adds a protocol if it's
+ * missing; urlencodes an anchor link; attempts to find the proper multilingual
+ * URL; and urlencodes any internal Backdrop path.
+ *
+ * @param string $url
+ *
+ * @return string
+ *   A cleaned-up URL that is either a properly formatted external URL or an
+ *   internal path.
+ *
+ * @see _filter_format_editor_link_url_validate()
+ */
+function _filter_cleanup_url($url) {
+  // Will add a default protocol if missing on an external url.
+  $url = _filter_create_valid_external_url($url);
+
+  if (_filter_url_is_external($url)) {
+    return $url;
+  }
+
+  // Urlencode an anchor as it's more likely to be typed in with unicode.
+  if (substr($url, 0, 1) == '#' && preg_match('/^(#)([\S]+)$/', $url, $matches)) {
+    return $matches['1'] . backdrop_encode_path($matches['2']);
+  }
+
+  // Break up the components so we can build up an aliased internal path.
+  $url_parts = backdrop_parse_url($url);
+
+  // Lookup valid internal paths if multilingual
+  if (module_exists('locale') && language_multilingual()) {
+    if ($language = _filter_lookup_locale_path($url)) {
+      $url_parts['language'] = $language;
+    }
+  }
+
+  return url($url_parts['path'], $url_parts);
+}
+
+/**
+ * Forms a valid external URL if possible.
+ *
+ * Trims whitespace and automatically adds an http:// to addresses that
+ * appear to be external URLs but have no specified protocol.
+ *
+ * @param string $url
+ *   The URL entered by the user.
+ *
+ * @return string
+ *   The URL with the protocol prepended if needed.
+ *
+ * @see link_cleanup_url().
+ */
+function _filter_create_valid_external_url($url) {
+  $url = trim($url);
+
+  // Check if there is no protocol specified.
+  $protocol_match = preg_match('/^([a-z0-9][a-z0-9\.\-_]*:\/\/)/i', $url);
+  if (empty($protocol_match)) {
+    // But should there be? Add an automatic http:// if it starts with a domain name.
+    $domain_match = preg_match('/^(([a-z0-9]([a-z0-9\-_]*\.){1,63})([a-z]{2,63}))/i', $url);
+    if (!empty($domain_match)) {
+      $url = "http://" . $url;
+    }
+  }
+
+  return $url;
+}
+
+/**
+ * Checks if a URL is external to Backdrop.
+ *
+ * Checks for a protocol as a sign that the URL is external. We want to include
+ * all external URLs even if they have dangerous protocols.
+ *
+ * @param string $url
+ *
+ * @return bool
+ *   TRUE if URL is external.
+ *
+ * @see url_is_external()
+ */
+function _filter_url_is_external($url) {
```

Comment:
Wouldn't `valid_url($url, TRUE)` just do the same thing?

Reply:
You might be right. I'll have to look into it.

Reply:
This didn't work. valid_url complains if the protocol is not one ftp, http, https, feed. If we do want to provide a warning of an invalid URL to the end user if they're using `javascript://boo` then `valid_url()` won't work.

Reply:
I don't quite understand... :confused: Why doesn't valid_url() work? It returns false and then we can warn, or what am I missing?